### ISSUE TITLE
How are external dependencies handled?

### ISSUE THREAD

**Issue by bhouston** [DelvarWorld/ShaderFrog-Runtime#2]
Let's say one had a shader file that is dependent upon textures or something like that?  How are those external dependencies handled?

Are they base64 inlined into the *.json file?

Are is the shader really a package of a set of files?

Or does shader execution rely on having access to the shaderFrog website for live loading of dependent assets?

**Comment by DelvarWorld** [comment]
At the moment it's assumed that any asset paths in the shader will be absolute URLs to the required assets. Actually, this is an oversight on my part, since that will mean all exported shaders with images are initially linked to my CDN.

This is open for discussions. It's under an assumption that the user might want to change an asset at runtime, like changing a texture. It's also an assumption that the end user is managing their own assets and has their own way to fetch binary resources like images. Base64 encoding in the shader itself would greatly increase the scope of the file format.

**Comment by bhouston** [comment]
Yeah, it is a challenging problem and I am not sure what the best solution is.

Clara.io hosts all of its assets on its own servers and we engage in deduplication to minimize load time.

We ideally would get a package of *.json per shader and then a set of images, or a JSON that allows us to fetch the images to load into the Clara CDN.  We would load the images into Clara.io as scene assets and then initialize a shader using the JSON.

Upon load of a *.json into a ThreeJS shader, it may be best to allow us to provide an optional function callback that can get an image request and return a texture asset.  This would allow us to repeatedly return the same asset if it is requested multiple times and it would allow us to use our own CDN.  It would also allow us to host that image asset on our servers and provide our image LOD capabilities.

I guess if each image has a CDN path and a UUID (and possibly a user friendly name or an image name), we can work with that.

I am not sure if base64 is required.  I guess I am working through this myself.
